pkg/sdkserver/localsdk_test.go

```diff
@@ -607,6 +610,130 @@ func TestLocalSDKServerPlayerConnectAndDisconnect(t *testing.T) {
+	l, err := NewLocalSDKServer(path, "")
+	assert.Nil(t, err)
+
+	stream := newGameServerMockStream()
+	go func() {
+		err := l.WatchGameServer(&sdk.Empty{}, stream)
+		assert.Nil(t, err)
+	}()
+	assertInitialWatchUpdate(t, stream)
+
+	// wait for watching to begin
+	err = wait.Poll(time.Second, 10*time.Second, func() (bool, error) {
+		found := false
+		l.updateObservers.Range(func(_, _ interface{}) bool {
+			found = true
+			return false
+		})
+		return found, nil
+	})
+	assert.NoError(t, err)
+
+	counterRequest := alpha.GetCounterRequest{Name: name}
+	expected := &alpha.Counter{Name: name, Count: &count, Capacity: &capacity}
+	counter, err := l.GetCounter(context.Background(), &counterRequest)
+	assert.NoError(t, err)
+	if diff := cmp.Diff(expected, counter, protocmp.Transform()); diff != "" {
+		t.Errorf("unexpected difference:\n%v", diff)
+	}
+}
+
+func TestLocalSDKServerUpdateCounter(t *testing.T) {
+	t.Parallel()
+
+	runtime.FeatureTestMutex.Lock()
+	defer runtime.FeatureTestMutex.Unlock()
+	assert.NoError(t, runtime.ParseFeatures(string(runtime.FeatureCountsAndLists)+"=true"))
+
+	name := "sessions"
+	counters := map[string]agonesv1.CounterStatus{
+		name: {Count: 1, Capacity: 100},
+	}
+	fixture := &agonesv1.GameServer{
+		ObjectMeta: metav1.ObjectMeta{Name: "stuff"},
+		Status: agonesv1.GameServerStatus{
+			Counters: counters,
+		},
+	}
+
+	path, err := gsToTmpFile(fixture)
+	assert.NoError(t, err)
+	l, err := NewLocalSDKServer(path, "")
+	assert.Nil(t, err)
+
+	stream := newGameServerMockStream()
+	go func() {
+		err := l.WatchGameServer(&sdk.Empty{}, stream)
+		assert.Nil(t, err)
+	}()
+	assertInitialWatchUpdate(t, stream)
+
+	// wait for watching to begin
+	err = wait.Poll(time.Second, 10*time.Second, func() (bool, error) {
+		found := false
+		l.updateObservers.Range(func(_, _ interface{}) bool {
+			found = true
+			return false
+		})
+		return found, nil
+	})
+	assert.NoError(t, err)
+
+	ONE := int64(1)
+	HUNDRED := int64(100)
+	notupdated := &alpha.Counter{Name: name, Count: &ONE, Capacity: &HUNDRED}
+	startingCounter, err := l.GetCounter(context.Background(), &alpha.GetCounterRequest{Name: "sessions"})
+	assert.NoError(t, err)
+	if diff := cmp.Diff(notupdated, startingCounter, protocmp.Transform()); diff != "" {
+		t.Errorf("unexpected difference:\n%v", diff)
+	}
+
+	count := int64(9)
+	capacity := int64(999)
+	// Note: the field mask filter in UpdateCounter will remove fields from the request object that
+	// are not included in the FieldMask Paths. Here `Name` and `Capacity` will be removed from
+	// the `updated` object after we call `l.UpdateCounter()`.
+	updated := &alpha.Counter{Name: name, Count: &count, Capacity: &capacity}
+	fm := fieldmaskpb.FieldMask{Paths: []string{"count"}}
+	expected := &alpha.Counter{Name: name, Count: &count, Capacity: &HUNDRED}
+	counterRequest := alpha.UpdateCounterRequest{Counter: updated, UpdateMask: &fm}
+	counter, err := l.UpdateCounter(context.Background(), &counterRequest)
```

Comment:
nit: more idiomatic would be `want` instead of `expected` and `got` instead of `counter`.

even nittier: this has more lvalues than I like, would generally inline `counterRequest` into `UpdateCounter` and I personally would end up inlining `fm` into the `UpdateCounterRequest` structure. I personally find this collection of variables a little difficult to read.